### PR TITLE
consensus and actpool to validate sidecar

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
as title

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/coreservice.go [hunk 1]
```diff
@@ -472,6 +472,10 @@ func (core *coreService) SendAction(ctx context.Context, in *iotextypes.Action)
 	if err != nil {
 		return "", err
 	}
+	ctx = protocol.WithBlockCtx(genesis.WithGenesisContext(ctx, core.Genesis()), protocol.BlockCtx{
+		ValidateSidecar: true,
```
Comment (dustinxie) [inline api/coreservice.go:476]: the other vars of `BlockCtx` are don't cares for actpool to add/validate the tx

File: consensus/scheme/rolldpos/rolldpos.go [hunk 2]
```diff
@@ -131,7 +131,7 @@ func (cm *chainManager) CommitBlock(blk *block.Block) error {
 
 // ValidateBlock validates a new block before adding it to the blockchain
 func (cm *chainManager) ValidateBlock(blk *block.Block) error {
-	return cm.bc.ValidateBlock(blk)
+	return cm.bc.ValidateBlock(blk, blockchain.ValidateSidecarOption())
```
Comment (dustinxie) [inline consensus/scheme/rolldpos/rolldpos.go:134]: consensus needs to validate blob sidecars (if present) for the proposed block

File: chainservice/builder.go [hunk 3]
```diff
@@ -539,8 +553,13 @@ func (builder *Builder) buildBlockSyncer() error {
 				retries = 4
 			}
 			var err error
+			opts := []blockchain.BlockValidationOption{}
+			if now := time.Now(); now.Before(blk.Timestamp()) ||
+				blk.Height()+cfg.Genesis.MinBlocksForBlobRequests > estimateHeight(blk, now.Sub(blk.Timestamp())) {
+				opts = append(opts, blockchain.ValidateSidecarOption())
+			}
```
Comment (dustinxie) [inline chainservice/builder.go:560]: we could go the other way, that is simply compare the timestamp, no need to bother with the `estimateHeight` above
```
if blk.Timestamp() <= time.Now() - 20 days {
    opts = append(opts, blockchain.ValidateSidecarOption())
}
```

File: chainservice/builder.go [hunk 4]
```diff
@@ -524,6 +524,20 @@ func (builder *Builder) buildBlockSyncer() error {
...
+	estimateHeight := func(blk *block.Block, duration time.Duration) uint64 {
+		if blk.Height() >= cfg.Genesis.DardanellesBlockHeight {
+			return blk.Height() + uint64(duration.Seconds()/float64(cfg.DardanellesUpgrade.BlockInterval))
+		}
+		durationToDardanelles := time.Duration(cfg.Genesis.DardanellesBlockHeight-blk.Height()) * time.Duration(cfg.Genesis.BlockInterval)
+		if duration < durationToDardanelles {
+			return blk.Height() + uint64(duration.Seconds()/float64(cfg.Genesis.BlockInterval))
+		}
+		return cfg.Genesis.DardanellesBlockHeight + uint64((duration-durationToDardanelles).Seconds()/float64(cfg.DardanellesUpgrade.BlockInterval))
+	}
```
Comment (dustinxie) [inline chainservice/builder.go:540]: no need to bother this, see comment below

File: blockchain/genesis/genesis.go [hunk 5]
```diff
@@ -43,42 +43,43 @@ func init() {
...
+			BlockGasLimit:            20000000,
+			TsunamiBlockGasLimit:     50000000,
+			ActionGasLimit:           5000000,
+			BlockInterval:            10 * time.Second,
+			NumSubEpochs:             2,
+			DardanellesNumSubEpochs:  30,
+			NumDelegates:             24,
+			NumCandidateDelegates:    36,
+			TimeBasedRotation:        false,
+			MinBlocksForBlobRequests: 345600,
```
Comment (CoderZhi) [inline blockchain/genesis/genesis.go:56]: this is more likely to be a derived value

File: chainservice/builder.go [hunk 6]
```diff
@@ -800,13 +819,13 @@ func (builder *Builder) build(forSubChain, forTest bool) (*ChainService, error)
...
-	if err := builder.buildBlockSyncer(); err != nil {
+	if err := builder.buildNodeInfoManager(); err != nil {
 		return nil, err
 	}
-	if err := builder.buildActionSyncer(); err != nil {
+	if err := builder.buildBlockSyncer(); err != nil {
 		return nil, err
 	}
-	if err := builder.buildNodeInfoManager(); err != nil {
+	if err := builder.buildActionSyncer(); err != nil {
```
Comment (CoderZhi) [inline chainservice/builder.go:828]: does the order matter?

File: action/protocol/generic_validator.go [hunk 7]
```diff
@@ -114,6 +114,12 @@ func (v *GenericValidator) Validate(ctx context.Context, selp *action.SealedEnve
 			if selp.BlobGasFeeCap().Cmp(basefee) < 0 {
 				return errors.Wrapf(action.ErrUnderpriced, "blob fee cap is too low: %s, base fee: %s", selp.BlobGasFeeCap().String(), basefee.String())
 			}
+			// validate sidecar
+			if blkCtx, ok := GetBlockCtx(ctx); ok && blkCtx.ValidateSidecar {
```
Comment (CoderZhi) [inline action/protocol/generic_validator.go:118]: if the sidecar does exist, but not validated, will it cause any problem?

File: action/protocol/generic_validator.go [hunk 8]
```diff
@@ -124,7 +124,7 @@ func (v *GenericValidator) Validate(ctx context.Context, selp *action.SealedEnve
 				return errors.Wrapf(action.ErrUnderpriced, "blob fee cap is too low: %s, base fee: %s", selp.BlobGasFeeCap().String(), basefee.String())
 			}
 			// validate sidecar
-			if blkCtx, ok := GetBlockCtx(ctx); ok && !blkCtx.SkipSidecarValidation {
+			if blkCtx, ok := GetBlockCtx(ctx); (ok && !blkCtx.SkipSidecarValidation) || selp.BlobTxSidecar() != nil {
```
Comment (dustinxie) [inline action/protocol/generic_validator.go:127]: if the tx contains a sidecar object, should always validate it

**Review by envestcc — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**